qutip/solver/floquet.py

```diff
@@ -0,0 +1,937 @@
+            ]
+
+    return [{k: out[k][i] for k in ks} for i in range(len(c_ops))]
+
+
+def floquet_gamma_matrices(X, delta, J_cb):
+    """
+    Floquet-Markov master equation gamma matrices.
+
+    Parameters
+    ----------
+    X : list of dict of :class:`qutip.data.Data`
+        Floquet delta tensor created by :func:`floquet_X_matrices`.
+
+    delta : np.ndarray
+        Floquet delta tensor created by :func:`floquet_delta_tensor`.
+
+    J_cb : list of callable
+        A list callback function that computes the noise power spectrum, as
+        a function of frequency, associated with each collapse operator `c_op`.
+        The function should take and return numpy arrays of frequencies.
+
+    Returns
+    -------
+    gammas : dict of :class:`qutip.data.Data`
+        A dict of the sidebands ``k`` to it's gamma matrices.
+    """
+    N = delta.shape[0]
+    kmax = (delta.shape[2] - 1) // 2
+    gamma = {k: _data.csr.zeros(N, N) for k in range(-kmax, kmax + 1, 1)}
+
+    for X_c_op, sp in zip(X, J_cb):
+        response = sp(delta) * ((2 + 0j) * np.pi)
+        response = [
+            _data.Dense(response[:, :, k], copy=False)
+            for k in range(2 * kmax + 1)
+        ]
+        for k in range(-kmax, kmax + 1, 1):
+            gamma[k] = _data.add(
+                gamma[k],
+                _data.multiply(
+                    _data.multiply(X_c_op[k].conj(), X_c_op[k]),
+                    response[k + kmax],
+                ),
+            )
+    return gamma
+
+
+def floquet_A_matrix(delta, gamma, w_th):
+    """
+    Floquet-Markov master equation rate matrix.
+
+    Parameters
+    ----------
+    delta : np.ndarray
+        Floquet delta tensor created by :func:`floquet_delta_tensor`.
+
+    gamma : dict of :class:`qutip.data.Data`
+        Floquet gamma matrices created by :func:`floquet_gamma_matrices`.
+
+    w_th : float
+        The temperature in units of frequency.
+    """
+    kmax = (delta.shape[2] - 1) // 2
+
+    if w_th > 0.0:
+        deltap = np.copy(delta)
+        deltap[deltap == 0.0] = np.inf
+        thermal = 1.0 / (np.exp(np.abs(deltap) / w_th) - 1.0)
+        thermal = [_data.Dense(thermal[:, :, k]) for k in range(2 * kmax + 1)]
+
+        gamma_kk = _data.add(gamma[0], gamma[0].transpose())
+        A = _data.add(gamma[0], _data.multiply(thermal[kmax], gamma_kk))
+
+        for k in range(1, kmax + 1):
+            g_kk = _data.add(gamma[k], gamma[-k].transpose())
+            thermal_kk = _data.multiply(thermal[kmax + k], g_kk)
+            A = _data.add(A, _data.add(gamma[k], thermal_kk))
+            thermal_kk = _data.multiply(thermal[kmax - k], g_kk.transpose())
+            A = _data.add(A, _data.add(gamma[-k], thermal_kk))
+    else:
+        # w_th is 0, thermal = 0s
+        A = gamma[0]
+        for k in range(1, kmax + 1):
+            A = _data.add(gamma[k], A)
+            A = _data.add(gamma[-k], A)
+
+    return A
+
+
+def floquet_master_equation_tensor(A):
```

Comment:
Let's give all of these internal `floquet_` methods a leading underscore. We can always make them public again later if someone wants them.